File: 1_two_sum.py
# https://leetcode.com/problems/two-sum/

class Solution(object):
    # O(n) time, O(n) space
    def twoSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        nums_to_indices = {}
        for i in range(len(nums)):
            num = nums[i]
            comp = target - num
            if comp in nums_to_indices:
                return [nums_to_indices[comp], i]
            nums_to_indices[num] = i
        return []

File: 1_TwoSum.cpp
// https://leetcode.com/problems/two-sum/

class Solution {
public:
    // O(n) time, O(n) space
    vector<int> twoSum(vector<int>& nums, int target) {
        unordered_map<int, int> num_to_idx;
        for (int i = 0; i < nums.size(); i++) {
            if (num_to_idx.count(target - nums[i]) != 0) {
                return {num_to_idx[target - nums[i]], i};
            }
            num_to_idx[nums[i]] = i;
        }
        return {};
    }
};
